keps/sig-node/2625-cpumanager-policies-thread-placement/README.md

```diff
@@ -0,0 +1,333 @@
+However, for some classes of these latency-sensitive applications running on simultaneous multithreading (SMT) enabled system, it is also beneficial
+to consider thread-level allocation, to avoid physical CPU sharing and prevent possible interferences caused by noisy neighborhoods.
+
+### Goals
+
+* Prevent workloads from requesting cores that don't consume a full CPU by rejecting them.
+  This guarantees that no physical core is shared among different containers, which improves cache efficiency and mitigates the noisy neighbours problem.
+
+## Proposal
+
+### User Stories
+
+#### Latency-sensitive applications runtime guarantees
+
+Some classes of latency-sensitive applications (CNF, HFT, ML/AI) benefit of the thread placement constraints this policy enables.
+More precise thread placement allows to control which physical cores are shared among containers, such as:
+
+1. workloads may ensure the physical cores are shared among their threads for increased efficiency
+2. workloads may ensure the physical cores are *not* shared with other containers for interference prevention.
+
+An implementation of the concepts proposed here is already found in [external projects](https://github.com/nokia/CPU-Pooler#hyperthreading-support).
+[OpenStack](https://specs.openstack.org/openstack/nova-specs/specs/mitaka/implemented/virt-driver-cpu-thread-pinning.html),
+which is one of the leading platform for VNF (Virtual Network Functions), the predecessor of CNFs.
+
+#### Improve the density of running containers
+
+Thread allocation guarantees enables more efficient usage of the node resources.
+The nodes can now accommodate safely mixed workloads of latency-sensitive and not-latency-sensitive (infrastructure) pods.
+This increases the node usage, which reduces the need for extra hardware, which drives down the TCO of a container-based solution.
+
+### Risks and Mitigations
+
+This new behaviour is opt-in. Users will need to explicitly enable it in their kubelet configuration. The change is very self contained, with little impact in the shared codebase.
+The impact in the shared codebase will be addressed enhancing the current testsuite.
+
+| Risk                                                      | Impact        | Mitigation |
+| --------------------------------------------------------- | ------------- | ---------- |
+| Bugs in the implementation lead to kubelet crash | High | Disable the policy and restart the kubelet. The workload will run but with weaker guarantees - like it was before this change. |
+
+
+## Design Details
+
+### Proposed Change
+
+We propose to add a new flag in Kubelet called `CPUManagerPolicyOptions` in the kubelet config or command line argument called `cpumanager-policy-options` which allows the user to specify the CPU Manager policy option. If the value of this option is specified to be `reject-non-smt-aligned`, it results in further refinements of the existing static policy.
```

Comment:
Frankly speaking, I'd prefer as separate policy name "smt-aligned-static" instead of option field to existing policy.
There are might be other components in the wild that reads kubelet config to determine node state and configured TopologyManager/CPUManager/... states. Instead of trying to teach those new field to parse (policy=static, option=xxx), might be better to have different policy names for different behaviours (policy=a vs. policy=b).

Reply:
We had a discussion on sig-node slack channel. The main issue of the `options` approach is that options can alter the behaviour of the `static` policy, but if legacy (= not updated) consumers/unaware operators looking only at the policy name can fail to detect this and run with wrong assumptions about the kubelet behaviour.
So, we will add a new policy name (`static-with-options`? `static-enhanced`?) which will be an alias of `static`, but will enable the consumption of the options. This means that to enable this new behaviour we are proposing, a operator would need to:
1. enable the feature gate (needed anyway)
2. change the policy name
3. actually provide options ( :) )

Reply:
See https://kubernetes.slack.com/archives/C0BP8PW9G/p1620737012440700 for more context